packages/@react-spectrum/tabs/src/Tabs.tsx

```diff
@@ -121,40 +180,230 @@ export function Tab<T>(props: TabProps<T>) {
+            ? <Text>{rendered}</Text>
+            : rendered}
+        </SlotProvider>
       </div>
     </FocusRing>
   );
 }
 
-function TabLine({orientation, selectedTab}) {
+interface TabLineProps {
+  orientation?: Orientation,
+  selectedTab?: HTMLElement,
+  selectedKey?: Key
+}
+
+function TabLine(props: TabLineProps) {
+  let {
+    orientation,
+    // Is either the tab node (non-collapsed) or the picker node (collapsed)
+    selectedTab,
+    // selectedKey is provided so that the TabLine styles are updated when the TabPicker's width updates from a selection change
+    selectedKey
+  } = props;
+
   let verticalSelectionIndicatorOffset = 12;
   let {direction} = useLocale();
   let {scale} = useProvider();
 
   let [style, setStyle] = useState({
     width: undefined,
-    height: undefined,
-    transform: undefined
+    height: undefined
   });
 
   useLayoutEffect(() => {
-    let styleObj = {transform: undefined, width: undefined, height: undefined};
-    styleObj.transform = orientation === 'vertical'
-      ? `translateY(${selectedTab.offsetTop + verticalSelectionIndicatorOffset / 2}px)`
-      : `translateX(${selectedTab.offsetLeft}px) `;
-
-    if (orientation === 'horizontal') {
-      styleObj.width = `${selectedTab.offsetWidth}px`;
-    } else {
-      styleObj.height = `${selectedTab.offsetHeight - verticalSelectionIndicatorOffset}px`;
+    if (selectedTab) {
+      let styleObj = {transform: undefined, width: undefined, height: undefined};
+      // In RTL, calculate the transform from the right edge of the tablist so that resizing the window doesn't break the Tabline position due to offsetLeft changes
+      let offset = direction === 'rtl' ? -1 * ((selectedTab.offsetParent as HTMLElement)?.offsetWidth - selectedTab.offsetWidth - selectedTab.offsetLeft) : selectedTab.offsetLeft;
+      styleObj.transform = orientation === 'vertical'
+        ? `translateY(${selectedTab.offsetTop + verticalSelectionIndicatorOffset / 2}px)`
+        : `translateX(${offset}px)`;
+
+      if (orientation === 'horizontal') {
+        styleObj.width = `${selectedTab.offsetWidth}px`;
+      } else {
+        styleObj.height = `${selectedTab.offsetHeight - verticalSelectionIndicatorOffset}px`;
+      }
+      setStyle(styleObj);
     }
-    setStyle(styleObj);
-  }, [direction, setStyle, selectedTab, orientation, scale]);
+
+  }, [direction, setStyle, selectedTab, orientation, scale, verticalSelectionIndicatorOffset, selectedKey]);
 
   return <div className={classNames(styles, 'spectrum-Tabs-selectionIndicator')} role="presentation" style={style} />;
 }
+
+interface CollapsibleTabListProps<T> extends TabListProps<T>, TabPickerProps<T> {
+  tabListProps?: HTMLAttributes<HTMLElement>,
+  wrapperRef: MutableRefObject<HTMLDivElement>,
+  collapse?: boolean,
+  tabListclassName?: string
+}
+
+const CollapsibleTabList = React.forwardRef(function <T> (props: CollapsibleTabListProps<T>, ref: MutableRefObject<HTMLDivElement>) {
```

Comment:
Originally it was in Tabs as well, but was split out like this for readability as per previous reviews. 

Since the CollapsibleTabList has a wrapping div (aka collapseWrapper) that shouldn't be present on a vertical tablist it looked something like this:
```
  let tablist = (
    <TabList
      {...tabListProps}
      ref={ref}
      orientation={orientation}
      density={density}
      isQuiet={isQuiet}
      isDisabled={isDisabled}
      state={state}
      selectedTab={selectedTab} />
  );

  if (collapse && orientation !== 'vertical') {
    tablist = (
      <TabPicker
        {...props}
        state={state} />
    );
  }

  if (orientation !== 'vertical') {
    tablist = (
      <div
        ref={wrapperRef}
        className={classNames(
          styles,
          'spectrum-Tabs--collapsible'
        )}>
        {tablist}
      </div>
    );
  }
```